### PR TITLE
[Clang] Workaround dependent source location issues

### LINKED ISSUE #80210: [18 regression] Assertion `!T->isDependentType() && "should not see dependent types here"' failed.
Test case extracted from qtwebengine 6.7.0-beta2:

```
$ clang++ -std=c++20 testcase.cpp
clang-18: /home/bero/abf/llvm/BUILD/llvm-project-release-18.x/build/tools/clang/include/clang/AST/TypeNodes.inc:36: TypeInfo clang::ASTContext::getTypeInfoImpl(const Type *) const: Assertion `!T->isDependentType() && "should not see dependent types here"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /usr/bin/clang-18 -cc1 -triple x86_64-pc-linux-gnu -emit-obj -mrelax-all -dumpdir a- -disable-free -clear-ast-before-backend -main-file-name testcase.cpp -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fdebug-compilation-dir=/tmp -fcoverage-compilation-dir=/tmp -resource-dir /usr/lib64/clang/18 -internal-isystem /usr/bin/../lib64/gcc/x86_64-openmandriva-linux-gnu/13.2.1/../../../../include/c++/13.2.1 -internal-isystem /usr/bin/../lib64/gcc/x86_64-openmandriva-linux-gnu/13.2.1/../../../../include/c++/13.2.1/x86_64-openmandriva-linux-gnu -internal-isystem /usr/bin/../lib64/gcc/x86_64-openmandriva-linux-gnu/13.2.1/../../../../include/c++/13.2.1/backward -internal-isystem /usr/lib64/clang/18/include -internal-isystem /usr/local/include -internal-isystem /usr/bin/../lib64/gcc/x86_64-openmandriva-linux-gnu/13.2.1/../../../../x86_64-openmandriva-linux-gnu/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -std=c++20 -fdeprecated-macro -ferror-limit 19 -fgnuc-version=13.1.1 -fno-implicit-modules -fcxx-exceptions -fexceptions -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/testcase-1cf133.o -x c++ testcase.cpp
1.      <eof> parser at end of file
2.      Per-file LLVM IR generation
3.      testcase.cpp:6:6: Generating code for declaration 'construct_at'
 #0 0x00007f0d0238e916 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib64/libLLVM-18.so+0x458e916)
 #1 0x00007f0d0238d58a llvm::sys::RunSignalHandlers() (/usr/lib64/libLLVM-18.so+0x458d58a)
 #2 0x00007f0d0238e234 (/usr/lib64/libLLVM-18.so+0x458e234)
 #3 0x00007f0cfd855700 (/usr/lib64/libc.so.6+0x3d700)
 #4 0x00007f0cfd8a49cc (/usr/lib64/libc.so.6+0x8c9cc)
 #5 0x00007f0cfd855652 gsignal (/usr/lib64/libc.so.6+0x3d652)
 #6 0x00007f0cfd83e4ad abort (/usr/lib64/libc.so.6+0x264ad)
 #7 0x00007f0cfd83e3d5 (/usr/lib64/libc.so.6+0x263d5)
 #8 0x00007f0cfd84e042 (/usr/lib64/libc.so.6+0x36042)
 #9 0x00007f0d093767ce clang::ASTContext::getTypeInfoImpl(clang::Type const*) const (/usr/lib64/libclang-cpp.so.18+0x19767ce)
#10 0x00007f0d09376f31 clang::ASTContext::getTypeInfo(clang::Type const*) const (/usr/lib64/libclang-cpp.so.18+0x1976f31)
#11 0x00007f0d0a9e4f42 clang::CodeGen::CodeGenFunction::CreateIRTemp(clang::QualType, llvm::Twine const&) (/usr/lib64/libclang-cpp.so.18+0x2fe4f42)
#12 0x00007f0d0ac29ffe clang::CodeGen::CodeGenFunction::StartFunction(clang::GlobalDecl, clang::QualType, llvm::Function*, clang::CodeGen::CGFunctionInfo const&, clang::CodeGen::FunctionArgList const&, clang::SourceLocation, clang::SourceLocation) (/usr/lib64/libclang-cpp.so.18+0x3229ffe)
#13 0x00007f0d0ac2cc99 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/usr/lib64/libclang-cpp.so.18+0x322cc99)
#14 0x00007f0d0ac770fe clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/usr/lib64/libclang-cpp.so.18+0x32770fe)
#15 0x00007f0d0ac764ca clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/usr/lib64/libclang-cpp.so.18+0x32764ca)
#16 0x00007f0d0ac4483f clang::CodeGen::CodeGenModule::EmitDeferred() (/usr/lib64/libclang-cpp.so.18+0x324483f)
#17 0x00007f0d0ac4486b clang::CodeGen::CodeGenModule::EmitDeferred() (/usr/lib64/libclang-cpp.so.18+0x324486b)
#18 0x00007f0d0ac4486b clang::CodeGen::CodeGenModule::EmitDeferred() (/usr/lib64/libclang-cpp.so.18+0x324486b)
#19 0x00007f0d0ac417ab clang::CodeGen::CodeGenModule::Release() (/usr/lib64/libclang-cpp.so.18+0x32417ab)
#20 0x00007f0d0ad0b364 (/usr/lib64/libclang-cpp.so.18+0x330b364)
#21 0x00007f0d0ac18830 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/usr/lib64/libclang-cpp.so.18+0x3218830)
#22 0x00007f0d09245416 clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib64/libclang-cpp.so.18+0x1845416)
#23 0x00007f0d0b77d34b clang::FrontendAction::Execute() (/usr/lib64/libclang-cpp.so.18+0x3d7d34b)
#24 0x00007f0d0b6e1799 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib64/libclang-cpp.so.18+0x3ce1799)
#25 0x00007f0d0b7fc71a clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib64/libclang-cpp.so.18+0x3dfc71a)
#26 0x0000560e96d84096 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/bin/clang-18+0x17096)
#27 0x0000560e96d7fd4f (/usr/bin/clang-18+0x12d4f)
#28 0x0000560e96d7e61b clang_main(int, char**, llvm::ToolContext const&) (/usr/bin/clang-18+0x1161b)
#29 0x0000560e96d8ef96 main (/usr/bin/clang-18+0x21f96)
#30 0x00007f0cfd83fc0a (/usr/lib64/libc.so.6+0x27c0a)
#31 0x00007f0cfd83fcc5 __libc_start_main (/usr/lib64/libc.so.6+0x27cc5)
#32 0x0000560e96d7bed1 _start (/usr/bin/clang-18+0xeed1)
```

Test case
```c++
struct BasicPersistent;
template <typename> BasicPersistent &&__declval(int);
template <typename _Tp> auto declval() -> decltype(__declval<_Tp>(0));
template <typename _Tp> _Tp forward;
template <typename _Tp, typename... _Args>
auto construct_at(_Tp *, _Args...) -> decltype(new _Tp(declval<_Args>()...)) {return 0;}
template <typename> struct allocator;
template <typename> struct allocator_traits;
template <typename _Tp> struct allocator_traits<allocator<_Tp>> {
  using pointer = _Tp *;
  template <typename _Up, typename... _Args>
  static void construct(_Up __p, _Args...) {
    construct_at(__p, forward<_Args>...);
  }
};
struct __alloc_traits : allocator_traits<allocator<BasicPersistent>> {
} push_back___x;
__alloc_traits::pointer _M_impl_0;
template <typename... _Args> void emplace_back(_Args...) {
  __alloc_traits::construct(_M_impl_0, forward<_Args>...);
}
struct SourceLocation {
  static SourceLocation Current(const char * = __builtin_FUNCTION());
};
struct BasicPersistent {
  BasicPersistent(BasicPersistent &&,
                  SourceLocation = SourceLocation::Current());
};
void CFXJSE_EngineAddObjectToUpArray() { emplace_back(push_back___x); }
```

[testcase.cpp.txt](https://github.com/llvm/llvm-project/files/14118017/testcase.cpp.txt)

### LINKED ISSUE #81155: source_location 18.x regression from 8c2b0d4175dcfe669a43d0173fd00ed3
This was found internally at Intel from some customer code. The original test case was stripped down to:
```cpp
template <bool> using enable_if_t = int;
template <class Ty> Ty Returns_exactly() noexcept;
struct InvokeBase {
  template <class Fx, class... Args>
  static auto Call(Fx func, Args... args) -> decltype(func(args...));
};
template <class Fx1, class Fx2 = Fx1, bool = false> struct Invoke;
template <class Fx1, class Fx2> struct Invoke<Fx1, Fx2, false> : InvokeBase {};
template <int> struct F { };
template <class...> struct InvocableR;
template <class Rx, class Callable, class... Args>
struct InvocableR<Rx, Callable, Args...> : 
  F<noexcept(Invoke<Rx>::Call(Returns_exactly<Rx>(), Returns_exactly<Callable>()))> {
  static constexpr bool value = false;
};
template <class Rx, class... Args> constexpr bool IsInvocableRV = InvocableR<Rx, Args...>::value;
template <class, class... Types> class FuncClass {
public:
  template <class Fx> using EnableIfCallable = enable_if_t<IsInvocableRV<Fx, Types...>>;
};
template <class> struct FuncImpl;
template <class Ret, class... Types> struct FuncImpl<Ret(Types...)> {
  using type = FuncClass<Ret, Types...>;
};
template <class FTy> class function {
  using Base = typename FuncImpl<FTy>::type;

public:
  template <class Fx, typename Base::template EnableIfCallable<Fx> = 0> function(Fx) {}
};

struct codeloc {
  static constexpr codeloc current_location(const char *name = __builtin_FUNCTION()) {
    return codeloc();
  }
  constexpr codeloc() {}
};
class buff {
public:
  buff(buff &, codeloc = codeloc::current_location());
};
void Help(function<void(buff)>) {
}
void Test() {
  Help([](buff) {});
}
```
https://godbolt.org/z/qao5jhv1e

And with help from @Endilll it was further reduced to:
```cpp
struct buff {
  buff(buff &, const char * = __builtin_FUNCTION());
};

template <class Ty>
Ty declval();

template <class Fx>
auto Call(buff arg) -> decltype(Fx{}(arg));

template <typename>
struct F {};

template <class Fx>
struct InvocableR : F<decltype(Call<Fx>(declval<buff>()))> {
  static constexpr bool value = false;
};

template <class Fx, bool = InvocableR<Fx>::value>
void Help(Fx) {}

void Test() {
  Help([](buff) {});
}
```
https://godbolt.org/z/3n3K46WT8

It's possibly worth noting that the behavior may be different on Windows than on Linux.

CC @cor3ntin

### LINKED ISSUE #85373: [clang][crash-on-valid] Crash in frontend with `std::source_location::current().column()` in deduction guide
The following code crashes Clang, compiling with `-std=c++20 -emit-llvm -Xclang -disable-llvm-passes`:

```cpp
#include <source_location>

template <int>
struct S { S(const char *); };

template <int M>
S(const char(&)[M]) -> S<std::source_location::current().column()>;

void f() {
    S{"If you remove this line, no crash happens."};
}
```

Link: https://godbolt.org/z/jPbb5Kr1P.
It does not crash on versions 17.1.0 and earlier.
I don't believe this is related to #67134, because this has a completely different backtrace.

It also gives an error:

```text
<source>:10:5: error: cannot compile this scalar expression yet
   10 |     S{"If you remove this line, no crash happens."};
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Backtrace:

```text
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-trunk/bin/clang++ -gdwarf-4 -g -o /app/output.s -mllvm --x86-asm-syntax=intel -S --gcc-toolchain=/opt/compiler-explorer/gcc-snapshot -fcolor-diagnostics -fno-crash-diagnostics -std=c++20 -emit-llvm -Xclang -disable-llvm-passes <source>
1.	<eof> parser at end of file
2.	<source>:9:6: LLVM IR generation of declaration 'f'
3.	<source>:9:6: Generating code for declaration 'f'
 #0 0x00000000036774c8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x36774c8)
 #1 0x0000000003675614 llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3675614)
 #2 0x00000000035bfc48 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007fbcd4242520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x0000000002f3a418 llvm::UndefValue::get(llvm::Type*) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x2f3a418)
 #5 0x0000000003db06e6 (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) CGExprScalar.cpp:0:0
 #6 0x0000000003db2a53 clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3db2a53)
 #7 0x0000000003d53f9e clang::CodeGen::CodeGenFunction::EmitAnyExpr(clang::Expr const*, clang::CodeGen::AggValueSlot, bool) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3d53f9e)
 #8 0x0000000003d6e83b clang::CodeGen::CodeGenFunction::EmitIgnoredExpr(clang::Expr const*) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3d6e83b)
 #9 0x000000000397f0be clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x397f0be)
#10 0x0000000003985319 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3985319)
#11 0x00000000039e2fc4 clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x39e2fc4)
#12 0x00000000039ef422 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x39ef422)
#13 0x0000000003a463bb clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3a463bb)
#14 0x0000000003a42105 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3a42105)
#15 0x0000000003a42db0 clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3a42db0)
#16 0x0000000003a4ae67 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (.part.0) CodeGenModule.cpp:0:0
#17 0x0000000003ecf3d6 (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) ModuleBuilder.cpp:0:0
#18 0x0000000003ec04f9 clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3ec04f9)
#19 0x0000000005d7496c clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x5d7496c)
#20 0x0000000003ecce75 clang::CodeGenAction::ExecuteAction() (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3ecce75)
#21 0x0000000004162361 clang::FrontendAction::Execute() (/opt/compiler-explorer/clang-trunk/bin/clang+++0x4162361)
#22 0x00000000040e831b clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x40e831b)
#23 0x00000000042449a3 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x42449a3)
#24 0x0000000000c32d06 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/compiler-explorer/clang-trunk/bin/clang+++0xc32d06)
#25 0x0000000000c2bb8d ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#26 0x0000000003f144f9 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#27 0x00000000035c0073 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x35c0073)
#28 0x0000000003f14719 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#29 0x0000000003edc457 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3edc457)
#30 0x0000000003edce0d clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3edce0d)
#31 0x0000000003ee6b6c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/opt/compiler-explorer/clang-trunk/bin/clang+++0x3ee6b6c)
#32 0x0000000000c30121 clang_main(int, char**, llvm::ToolContext const&) (/opt/compiler-explorer/clang-trunk/bin/clang+++0xc30121)
#33 0x0000000000b26194 main (/opt/compiler-explorer/clang-trunk/bin/clang+++0xb26194)
#34 0x00007fbcd4229d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#35 0x00007fbcd4229e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#36 0x0000000000c2b36e _start (/opt/compiler-explorer/clang-trunk/bin/clang+++0xc2b36e)
```

### LINKED ISSUE #106428: [libc++] std::invoke() substitution failure when using function object with default argument
The following test program compiles with libc++17 and libstdc++, but fails with libc++18:

```cpp
#include <concepts>
#include <functional>
#include <ranges>
#include <source_location>

#define FWD(x) static_cast<decltype(x)&&>(x)

struct add_fn {
    template <std::integral T>
    constexpr auto operator()(T lhs, T rhs,
                              std::source_location loc = std::source_location::current())
        const -> T
    {
        return lhs + rhs;
    }
};

inline constexpr auto add = add_fn{};

struct fold_op {
    template <typename Rng, typename Func, typename Init>
    constexpr auto operator()(Rng&& rng, Func func, Init init) const
    {
        auto init_ = std::ranges::range_value_t<Rng>(std::move(init));
        for (auto&& elem : rng) {
            init_ = std::invoke(func, std::move(init_), FWD(elem));
        }
        return init_;
    }
};

inline constexpr auto fold = fold_op{};


auto sum(std::span<int const> arr) -> int
{
    return fold(arr, add, 0);
}
```

https://godbolt.org/z/h47dnM3eP

The error is:

```
<source>:27:21: error: no matching function for call to 'invoke'
   27 |             init_ = std::invoke(func, std::move(init_), FWD(elem));
      |                     ^~~~~~~~~~~
<source>:38:16: note: in instantiation of function template specialization 'fold_op::operator()<std::span<const int> &, add_fn, int>' requested here
   38 |     return fold(arr, add, 0);
      |                ^
/opt/compiler-explorer/clang-18.1.0/bin/../include/c++/v1/__functional/invoke.h:27:1: note: candidate template ignored: substitution failure [with _Fn = add_fn &, _Args = <__libcpp_remove_reference_t<int &>, const int &>]
   27 | invoke(_Fn&& __f, _Args&&... __args) noexcept(is_nothrow_invocable_v<_Fn, _Args...>) {
      | ^
1 error generated.
Compiler returned: 1
```

Removing the `source_location` default argument from `add_fn::operator()` allows the code to compile as expected.

### PR DESCRIPTION
**Pull request by cor3ntin**
In #78436 we made some SourceLocExpr dependent to
deal with the fact that their value should reflect the name of specialized function - rather than the rtemplate in which they are first used.

However SourceLocExpr are unusual in two ways
 - They don't depend on template arguments
 - They morally depend on the context in which they are used (rather than called from).

It's fair to say that this is quite novels and confuses clang. In particular, in some cases, we used to create dependent SourceLocExpr and never subsequently transform them, leaving dependent objects in instantiated functions types. To work around that we avoid replacing SourceLocExpr when we think they could remain dependent.
It's certainly not perfect but it fixes a number of reported bugs, and seem to only affect scenarios in which the value of the SourceLocExpr does not matter (overload resolution).

Fixes #106428
Fixes #81155
Fixes #80210
Fixes #85373

### CODE REVIEW

**Comment by zyn0217** [comment]
Does it fix #80210?

**Comment by cor3ntin** [comment]
@AaronBallman btw, backporting that would be nice...

File: clang/lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -5443,11 +5443,24 @@ struct EnsureImmediateInvocationInDefaultArgs
...
       return E;
-    return getDerived().RebuildSourceLocExpr(E->getIdentKind(), E->getType(),
-                                             E->getBeginLoc(), E->getEndLoc(),
-                                             SemaRef.CurContext);
+
+    // FIXME: During instantiation, because the rebuild of defaults arguments
+    // is not always done in the context of the template instantiator,
+    // we run the risk of producing a dependent source location
+    // that would never be rebuilt.
+    // This usually happen during overload resolution, or in contexts
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:5454]:
```suggestion
    // This usually happens during overload resolution, or in contexts
```

File: clang/lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -5443,11 +5443,24 @@ struct EnsureImmediateInvocationInDefaultArgs
...
+    // where the value of the source location does not matter.
+    // However, we should find a better way to deal with source location
+    // of function template.
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:5457]:
```suggestion
    // of function templates.
```

**Review by AaronBallman — APPROVED** [review]
LGTM with some minor grammatical nits in a comment.

File: clang/lib/Sema/SemaExpr.cpp [hunk 3]
```diff
@@ -5443,11 +5443,24 @@ struct EnsureImmediateInvocationInDefaultArgs
...
+    // FIXME: During instantiation, because the rebuild of defaults arguments
+    // is not always done in the context of the template instantiator,
+    // we run the risk of producing a dependent source location
+    // that would never be rebuilt.
+    // This usually happens during overload resolution, or in contexts
+    // where the value of the source location does not matter.
+    // However, we should find a better way to deal with source location
+    // of function templates.
+    if (!SemaRef.CurrentInstantiationScope ||
+        !SemaRef.CurContext->isDependentContext() || DC->isDependentContext())
```
Comment (shafik) [inline clang/lib/Sema/SemaExpr.cpp:5459]: Are we sure the `DC` is always not a `nullptr`?

Comment (cor3ntin) [reply clang/lib/Sema/SemaExpr.cpp:5459]: Yes, there is always a declaration context (the top one being that of the translation unit)

**Pull request merged**